File: scripts/train-digits.mjs
/**
 * Build and measure the digit classifier.
 *
 * Approach: nearest-neighbour over normalized 28x28 bitmaps. Not a CNN, and
 * deliberately so -- there is no ML toolchain on this machine, and with a few
 * thousand training digits a nearest-neighbour model is competitive while being
 * inspectable: every prediction can be traced to the specific training digit it
 * matched, which matters when the whole design rests on knowing when NOT to
 * trust a reading.
 *
 * ---------------------------------------------------------------------------
 * What this is FOR, which decides how it is measured.
 *
 * The recognizer does not replace the human. It pre-fills a box that a person
 * is looking at anyway, beside a picture of the handwriting. So the number that
 * matters is not accuracy over all digits -- it is PRECISION ON THE ONES IT
 * CHOOSES TO ANSWER. A cell it declines to fill costs a keystroke the human was
 * going to make regardless. A cell it fills wrongly costs far more, because it
 * invites agreement, and a wrong number that looks confident is the one failure
 * this project is arranged to avoid.
 *
 * So the output below is a precision/coverage curve, not a single accuracy
 * figure, and the threshold is chosen for precision with coverage as whatever
 * falls out.
 * ---------------------------------------------------------------------------
 *
 * Measured leave-one-SCAN-out: the model is trained on 17 events and tested on
 * the 18th, then rotated. Anything less flatters the result -- digits from one
 * event share a handful of volunteers, one pen each, and one scanner session,
 * so a digit from the same card (or even the same event) is a much easier test
 * than the real one, which is a new event by new people next month.
 *
 * Labels come from `label-from-spreadsheet.mjs`; see out/training/.
 *
 * Usage:
 *   node scripts/train-digits.mjs            measure
 *   node scripts/train-digits.mjs --emit     also write the model
 */

import { readdirSync, readFileSync, writeFileSync, mkdirSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

import { distance, prepare, shiftVariants, unitNorm } from "../src/lib/digits.ts";

const ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");
const TRAINING = join(ROOT, "out", "training");
const REF = join(ROOT, "assets", "reference");

/** Neighbours polled per prediction. */
const K = 5;

/**
 * Scale a bitmap to unit length.
 *
 * Without this, L2 distance is partly a comparison of how much ink each digit
 * carries, so a heavily-written 1 can sit closer to a 0 than to a light 1.
 * Normalizing leaves only the shape.
 */
/**
 * The preparation and the distance both live in src/lib/digits.ts now, so the
 * browser and this measurement run the same code. A query prepared any
 * differently from the exemplars is comparing nothing, which is exactly the
 * kind of drift a second copy invites.
 */

export function loadTrainingSet() {
  const samples = [];
  for (const file of readdirSync(TRAINING).filter((f) => f.endsWith(".json"))) {
    const data = JSON.parse(readFileSync(join(TRAINING, file), "utf8"));
    for (const s of data.samples ?? []) {
      samples.push({
        label: s.label,
        bitmap: prepare(s.bitmap),
        // Kept unprepared for --emit: the shipped model stores raw 0-255
        // bytes and the app prepares them, which is a third of the size.
        raw: Uint8Array.from(s.bitmap),
        source: s.source ?? data.source ?? file.replace(/\.json$/, ""),
        // A cell is identified by scan + card + row; its digits stand or fall
        // together when a value is judged.
        cell: `${s.source}:${s.card}:${s.row}`,
        value: s.value,
      });
    }
  }
  return samples;
}

/**
 * Classify by polling the K nearest training digits.
 *
 * Confidence is the share of the poll won by the top class, weighted by
 * closeness. It is what the pre-fill is gated on, so it is checked below that
 * it actually tracks correctness rather than merely correlating with it.
 *
 * The pool size, the shift re-score and the distance all come from
 * src/lib/digits.ts, which is what the browser runs.
 */
const POOL = 25;

function classify(bitmap, train, k = K) {
  const pool = [];
  for (const s of train) {
    const cutoff = pool.length < POOL ? Infinity : pool[pool.length - 1].d;
    const d = distance(bitmap, s.bitmap, cutoff);
    if (d === Infinity) continue;

    pool.push({ d, label: s.label, bitmap: s.bitmap });
    pool.sort((x, y) => x.d - y.d);
    if (pool.length > POOL) pool.pop();
  }
  if (pool.length === 0) return { label: null, confidence: 0 };

  // Re-score the pool allowing the query to move, then keep the k best.
  const variants = shiftVariants(bitmap);
  const best = pool
    .map((c) => {
      let m = Infinity;
      for (const v of variants) {
        const d = distance(v, c.bitmap, m);
        if (d < m) m = d;
      }
      return { d: m, label: c.label };
    })
    .sort((x, y) => x.d - y.d)
    .slice(0, k);

  const weights = new Map();
  let total = 0;
  for (const b of best) {
    // Inverse-distance weighting, so a very close match counts for more than a
    // distant one that happens to be inside the poll.
    const w = 1 / (Math.sqrt(b.d) + 1e-6);
    weights.set(b.label, (weights.get(b.label) ?? 0) + w);
    total += w;
  }

  let label = null;
  let bestW = -1;
  for (const [l, w] of weights) {
    if (w > bestW) {
      bestW = w;
      label = l;
    }
  }
  return { label, confidence: total ? bestW / total : 0 };
}

function main() {
  const samples = loadTrainingSet();
  const scans = [...new Set(samples.map((s) => s.source))];

  console.log(`training digits: ${samples.length} from ${scans.length} events`);
  const perClass = new Array(10).fill(0);
  for (const s of samples) perClass[s.label]++;
  console.log(`per class: ${perClass.map((n, i) => `${i}:${n}`).join("  ")}`);

  // ---- leave-one-scan-out ----
  const results = [];
  for (const held of scans) {
    const train = samples.filter((s) => s.source !== held);
    const test = samples.filter((s) => s.source === held);
    for (const s of test) {
      const { label, confidence } = classify(s.bitmap, train);
      results.push({ ...s, predicted: label, confidence, correct: label === s.label });
    }
    process.stdout.write(`\r  tested ${results.length}/${samples.length}   `);
  }
  process.stdout.write("\n");

  const acc = results.filter((r) => r.correct).length / results.length;
  console.log(`\nper-digit accuracy, all digits: ${(acc * 100).toFixed(1)}%`);

  // ---- the number that decides the design: precision where it answers ----
  console.log("\nprecision/coverage by confidence threshold:");
  console.log("  threshold   answered   precision   cells fully right");
  for (const t of [0, 0.5, 0.6, 0.7, 0.8, 0.9, 0.95, 1.0]) {
    const answered = results.filter((r) => r.confidence >= t);
    if (answered.length === 0) continue;
    const right = answered.filter((r) => r.correct).length;

    // A value is only usable if EVERY digit in it was answered and right --
    // "2" and "21" are different numbers of debris.
    const byCell = new Map();
    for (const r of results) {
      if (!byCell.has(r.cell)) byCell.set(r.cell, []);
      byCell.get(r.cell).push(r);
    }
    let fullCells = 0;
    let wholeCells = 0;
    for (const digits of byCell.values()) {
      if (!digits.every((d) => d.confidence >= t)) continue;
      wholeCells++;
      if (digits.every((d) => d.correct)) fullCells++;
    }

    console.log(
      `  >= ${t.toFixed(2)}    ` +
        `${String(answered.length).padStart(5)} (${String(Math.round((answered.length / results.length) * 100)).padStart(3)}%)   ` +
        `${((right / answered.length) * 100).toFixed(1)}%      ` +
        `${wholeCells ? ((fullCells / wholeCells) * 100).toFixed(1) : "-"}% of ${wholeCells} cells`,
    );
  }

  // ---- what it gets wrong, so the failures are known rather than assumed ----
  const confusion = Array.from({ length: 10 }, () => new Array(10).fill(0));
  for (const r of results) if (r.predicted !== null) confusion[r.label][r.predicted]++;

  console.log("\nconfusion (row = truth, col = predicted):");
  console.log("     " + [...Array(10).keys()].map((i) => String(i).padStart(5)).join(""));
  confusion.forEach((row, i) => {
    console.log(`  ${i}: ` + row.map((n) => String(n || "").padStart(5)).join(""));
  });

  const pairs = [];
  confusion.forEach((row, t) =>
    row.forEach((n, p) => {
      if (t !== p && n > 0) pairs.push({ t, p, n });
    }),
  );
  pairs.sort((a, b) => b.n - a.n);
  console.log("\nworst confusions:");
  for (const { t, p, n } of pairs.slice(0, 8)) console.log(`  ${t} read as ${p}: ${n}`);

  console.log("\nper class recall:");
  for (let d = 0; d < 10; d++) {
    const of = results.filter((r) => r.label === d);
    if (!of.length) continue;
    const right = of.filter((r) => r.correct).length;
    console.log(
      `  ${d}: ${String(right).padStart(4)}/${String(of.length).padEnd(4)} ` +
        `${((right / of.length) * 100).toFixed(0)}%`,
    );
  }

  console.log("\nper event (a whole unseen event each time):");
  for (const scan of scans) {
    const of = results.filter((r) => r.source === scan);
    const right = of.filter((r) => r.correct).length;
    console.log(
      `  ${scan.padEnd(16)} ${String(right).padStart(4)}/${String(of.length).padEnd(4)} ` +
        `${((right / of.length) * 100).toFixed(0)}%`,
    );
  }

  if (process.argv.includes("--emit")) {
    mkdirSync(REF, { recursive: true });
    const model = {
      kind: "knn-28x28",
      k: K,
      note:
        "Nearest-neighbour digit model. Bitmaps are 28x28, ink 0-255, scaled " +
        "to fit 20x20 and centred by centre of mass (MNIST convention). " +
        "Exemplars are RAW bytes, base64: run prepare() from src/lib/digits.ts " +
        "over each one before comparing, and over the query too, or the " +
        "distances mean nothing. At read time the query is also tried at nine " +
        "one-pixel offsets and the closest is kept.",
      digitAccuracy: Number(acc.toFixed(4)),
      trainedOn: scans,
      // Raw bytes, base64. Storing the PREPARED float vectors instead costs
      // 8.1MB against 3.5MB, and the preparation is a few milliseconds over the
      // whole set at load time.
      samples: samples.map((s) => ({
        label: s.label,
        b: Buffer.from(s.raw).toString("base64"),
      })),
    };
    const path = join(REF, "digit-model.json");
    writeFileSync(path, JSON.stringify(model));
    console.log(`\nwrote ${path} (${samples.length} exemplars)`);
  }
}

// Guarded so the preparation helpers above can be imported by the rule sweep
// without this file measuring anything on the way in.
if (import.meta.url === `file://${process.argv[1]}`) main();

File: scripts/train_digits.py
"""Train and measure a digit classifier that can be shipped to the browser.

The nearest-neighbour model in ``train-digits.mjs`` saturated: 7, 18 and 27
events gave 63.0%, 66.3% and 66.3%, so more data stopped helping, and the
precision where it answers never moved off ~84%.  A note there says there is no
ML toolchain on this machine.  That is no longer true -- ``pip install --target``
puts scikit-learn on this laptop in about a minute -- so this is the honest next
method.

Two things are being tested at once, and the second matters more than the first:

*A model with a decision boundary* rather than a poll of neighbours.  Nearest
neighbour over raw pixels asks whether this digit looks like one it has seen,
which is a question about the training set; a fitted model asks what separates
the classes, which is a question about digits.

*Augmentation.*  There are 3,218 digits and they are wildly unbalanced -- 1,142
ones against 50 nines -- because that is how counts of beach debris are
distributed, and the chapter cannot be asked to write more nines.  What can be
done is to make more of the ones it already has: the same digit shifted a pixel,
turned a couple of degrees, or written slightly larger is a different bitmap and
the same number.  This is the one lever the earlier work did not pull, and on a
set this small it is usually worth more than the choice of model.

Measured leave-one-EVENT-out, like everything else here: train on 26 events,
test on the 27th, rotate.  Anything less flatters the result, because digits
from one event share a handful of volunteers, one pen each and one scanner
session.

The figure that decides the design is PRECISION WHERE IT ANSWERS, not accuracy.
A cell the model declines costs a keystroke the human was making anyway; a cell
it fills wrongly costs data integrity, because a confident wrong number invites
agreement.

Usage:
    PYTHONPATH=<dir-with-sklearn> python3 scripts/train_digits.py [--emit]
"""

from __future__ import annotations

import argparse
import json
import pathlib
import sys

import numpy as np

ROOT = pathlib.Path(__file__).resolve().parent.parent
TRAINING = ROOT / "out" / "training"
REF = ROOT / "assets" / "reference"

SIDE = 28


def load() -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """Return bitmaps, labels, source event, and cell id for every digit."""
    bitmaps, labels, sources, cells = [], [], [], []
    for path in sorted(TRAINING.glob("*.json")):
        data = json.loads(path.read_text())
        for sample in data.get("samples", []):
            bitmaps.append(sample["bitmap"])
            labels.append(sample["label"])
            source = sample.get("source") or data.get("source") or path.stem
            sources.append(source)
            # A value stands or falls as a whole: "2" and "21" are different
            # numbers of debris, so a cell is only usable if every digit in it
            # was answered and answered right.
            cells.append(f"{source}:{sample['card']}:{sample['row']}")
    return (
        # float64, not float32. In float32 the optimiser's own accumulators
        # overflow -- numpy reports it in the matmuls -- and a diverged net still
        # returns a softmax that looks exactly like a confidence, which is the
        # worst possible failure for a design that gates on confidence.
        np.asarray(bitmaps, dtype=np.float64) / 255.0,
        np.asarray(labels, dtype=np.int64),
        np.asarray(sources),
        np.asarray(cells),
    )


def shift(img: np.ndarray, dx: int, dy: int) -> np.ndarray:
    out = np.zeros_like(img)
    xs = slice(max(0, dx), SIDE + min(0, dx))
    ys = slice(max(0, dy), SIDE + min(0, dy))
    xd = slice(max(0, -dx), SIDE + min(0, -dx))
    yd = slice(max(0, -dy), SIDE + min(0, -dy))
    out[ys, xs] = img[yd, xd]
    return out


def rotate(img: np.ndarray, degrees: float) -> np.ndarray:
    """Rotate about the centre with nearest-neighbour sampling.

    Nearest neighbour rather than anything smoother on purpose: these bitmaps
    are ink COVERAGE, already box-filtered down from the scan, and interpolating
    them again blurs away the thin strokes that separate a 1 from a 7.
    """
    theta = np.deg2rad(degrees)
    cos, sin = np.cos(theta), np.sin(theta)
    ys, xs = np.mgrid[0:SIDE, 0:SIDE]
    cx = cy = (SIDE - 1) / 2
    sx = (xs - cx) * cos + (ys - cy) * sin + cx
    sy = -(xs - cx) * sin + (ys - cy) * cos + cy
    sx = np.rint(sx).astype(int)
    sy = np.rint(sy).astype(int)
    ok = (sx >= 0) & (sx < SIDE) & (sy >= 0) & (sy < SIDE)
    out = np.zeros((SIDE, SIDE), dtype=img.dtype)
    out[ys[ok], xs[ok]] = img[sy[ok], sx[ok]]
    return out


def scale(img: np.ndarray, factor: float) -> np.ndarray:
    ys, xs = np.mgrid[0:SIDE, 0:SIDE]
    c = (SIDE - 1) / 2
    sx = np.rint((xs - c) / factor + c).astype(int)
    sy = np.rint((ys - c) / factor + c).astype(int)
    ok = (sx >= 0) & (sx < SIDE) & (sy >= 0) & (sy < SIDE)
    out = np.zeros((SIDE, SIDE), dtype=img.dtype)
    out[ys[ok], xs[ok]] = img[sy[ok], sx[ok]]
    return out


def augment(x: np.ndarray, y: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """Every digit, plus small shifts, rotations and rescalings of it.

    Deliberately small.  These are variations one hand produces writing the same
    digit twice, which is the variation the model has to survive; anything
    larger starts inventing digits nobody wrote.
    """
    imgs = x.reshape(-1, SIDE, SIDE)
    out = [imgs]
    out.append(np.stack([shift(i, 1, 0) for i in imgs]))
    out.append(np.stack([shift(i, -1, 0) for i in imgs]))
    out.append(np.stack([shift(i, 0, 1) for i in imgs]))
    out.append(np.stack([shift(i, 0, -1) for i in imgs]))
    out.append(np.stack([rotate(i, 8) for i in imgs]))
    out.append(np.stack([rotate(i, -8) for i in imgs]))
    out.append(np.stack([scale(i, 1.12) for i in imgs]))
    out.append(np.stack([scale(i, 0.89) for i in imgs]))
    xs = np.concatenate(out).reshape(-1, SIDE * SIDE)
    ys = np.tile(y, len(out))
    return xs, ys


def build(seed: int = 0):
    from sklearn.neural_network import MLPClassifier
    from sklearn.pipeline import make_pipeline
    from sklearn.preprocessing import StandardScaler

    # Trained gently and stopped early, on purpose. The first version ran at
    # 3e-3 for a fixed sixty epochs and diverged -- numpy reported overflow in
    # the matmuls -- and a diverged net still returns a softmax that looks like
    # a confidence, which is the worst possible failure for a design that gates
    # on confidence.
    # Standardized first, and that is not decoration. These bitmaps are ink
    # COVERAGE: most of the 784 inputs are zero for every digit in the set,
    # because no digit reaches the corners of its box. Feeding columns with no
    # variance straight into an optimiser is what made the first two attempts
    # diverge -- numpy reported overflow in the matmuls, and a diverged network
    # still returns a softmax that looks exactly like a confidence, which is the
    # worst possible failure for a design that gates on confidence.
    return make_pipeline(
        StandardScaler(),
        MLPClassifier(
            hidden_layer_sizes=(200,),
            alpha=1e-3,
            batch_size=128,
            learning_rate_init=3e-4,
            max_iter=300,
            early_stopping=True,
            n_iter_no_change=12,
            random_state=seed,
        ),
    )


def curve(pred: np.ndarray, conf: np.ndarray, truth: np.ndarray, cells: np.ndarray) -> None:
    print("\n  threshold   answered      precision   whole cells right")
    for t in (0.0, 0.5, 0.7, 0.9, 0.95, 0.99, 0.995):
        mask = conf >= t
        if not mask.any():
            continue
        precision = (pred[mask] == truth[mask]).mean()

        # A cell counts only when every digit in it was answered.
        whole = right = 0
        for cell in np.unique(cells):
            at = cells == cell
            if not mask[at].all():
                continue
            whole += 1
            right += bool((pred[at] == truth[at]).all())
        print(
            f"  >= {t:.3f}   {mask.sum():5d} ({mask.mean() * 100:3.0f}%)   "
            f"{precision * 100:7.1f}%   "
            f"{(right / whole * 100) if whole else 0:6.1f}% of {whole}"
        )


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--emit", action="store_true", help="write the model for the browser")
    parser.add_argument("--no-augment", action="store_true")
    args = parser.parse_args()

    x, y, source, cells = load()
    events = np.unique(source)
    print(f"{len(y)} digits from {len(events)} events")
    print("per class: " + "  ".join(f"{d}:{(y == d).sum()}" for d in range(10)))

    pred = np.zeros_like(y)
    conf = np.zeros(len(y), dtype=np.float64)

    for i, held in enumerate(events, 1):
        train = source != held
        xt, yt = x[train], y[train]
        if not args.no_augment:
            xt, yt = augment(xt, yt)
        model = build()
        model.fit(xt, yt)
        probs = model.predict_proba(x[~train])
        pred[~train] = model.classes_[probs.argmax(axis=1)]
        conf[~train] = probs.max(axis=1)
        print(f"\r  fold {i}/{len(events)}  ({held})            ", end="", flush=True)
    print()

    print(f"\nper-digit accuracy, all digits: {(pred == y).mean() * 100:.1f}%")
    curve(pred, conf, y, cells)

    print("\nper class recall:")
    for d in range(10):
        at = y == d
        if at.any():
            hit, n = (pred[at] == d).sum(), at.sum()
            print(f"  {d}: {hit:4d}/{n:<4d} {hit / n * 100:3.0f}%")

    if args.emit:
        xt, yt = (x, y) if args.no_augment else augment(x, y)
        model = build()
        model.fit(xt, yt)
        blob = {
            "kind": "mlp-28x28",
            "note": (
                "Bitmaps are 28x28 ink coverage 0-255, scaled to fit 20x20 and "
                "centred by centre of mass. Inputs are divided by 255. Layers "
                "are dense with relu, then softmax."
            ),
            "layers": [
                {"w": w.tolist(), "b": b.tolist()}
                for w, b in zip(model[-1].coefs_, model[-1].intercepts_)
            ],
            "scaler": {
                "mean": model[0].mean_.tolist(),
                "scale": model[0].scale_.tolist(),
            },
            "classes": model[-1].classes_.tolist(),
        }
        path = REF / "digit-model.json"
        path.write_text(json.dumps(blob))
        print(f"\nwrote {path} ({path.stat().st_size // 1024} KB)")
    return 0


if __name__ == "__main__":
    sys.exit(main())
